internal/client/db/db.go

```diff
@@ -0,0 +1,585 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package db
+
+import (
+	"fmt"
+	"log"
+	"sync"
+
+	"github.com/ChainSafe/gossamer/internal/client/api"
+	"github.com/ChainSafe/gossamer/internal/client/db/columns"
+	"github.com/ChainSafe/gossamer/internal/client/db/metakeys"
+	"github.com/ChainSafe/gossamer/internal/primitives/blockchain"
+	"github.com/ChainSafe/gossamer/internal/primitives/core/hash"
+	"github.com/ChainSafe/gossamer/internal/primitives/database"
+	"github.com/ChainSafe/gossamer/internal/primitives/runtime"
+	"github.com/ChainSafe/gossamer/internal/primitives/runtime/generic"
+	"github.com/ChainSafe/gossamer/pkg/scale"
+	"github.com/li1234yun/gods-generic/maps/linkedhashmap"
+)
+
+const numCachedHeaders = 8
+
+// An extrinsic entry in the database.
+type dbExtrinsic[E runtime.Extrinsic] struct {
```

Comment:
Can you please explan, why do we need generic here since runtime.Extrinsic is an interface?

Reply:
I think it makes it clear that one or multiple types that will be stored in `inner` will require an implementation of `runtime.Extrinsic`.  Is it absolutely necessary though, no it's not.

Reply:
> Is it absolutely necessary though, no it's not
Sorry I do not understand this, probably artistic turns of phrase.
But if I understand correctly, idea is that in case you create  instance of dbExtrinsic you wan to be sure that you are always have to pass the same type once you initiated a structure? 
What are the occasions that this might occurs?

Reply:
Sorry I misspoke.  I use the generic param `runtime.Extrinsic` to be used in the functions that implement `scale.VaryingDataType`.  This is essentially used to populate `dbExtrinsic.inner` with the right extrinsic type for `dbExtrinsicFull`.  Receiver functions need the generic param definitions associated with the type.  You can't introduce new generic params in receiver functions.

Reply:
Ok, I misspoke again.  The `dbExtrinsicFull` type is actually SCALE encoded and decoded into.  By using a generic param for `runtime.Extrinsic` it allows us to specify the destination type in the struct instead of treat it as an interface.  FYI, in SCALE we can encode an interface, but we can't decode into an interface given that we don't know the underlying type is that implements the interface.